// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package steeleagle.protocol.common;

import "google/protobuf/timestamp.proto";

// Request object for additional request info.
message Request {
  google.protobuf.Timestamp timestamp = 1; // request timestamp
}

/*
 * Response types for RPC functions.
 *
 * Values 0-2 (`OK`, `IN_PROGRESS`, `COMPLETED`) are specific to the SteelEagle protocol.
 * They determine what phase an RPC call is in:
 * - `OK` -> ack
 * - `IN_PROGRESS` -> in progress
 * - `COMPLETED` -> completed
 *
 * These intermediate phases are generally not exposed to user-facing code and are only used 
 * for streaming methods. In contrast to normal gRPC procedure, a call is only considered complete
 * on a response of `COMPLETED` instead of `OK`. If an error occurs, SteelEagle defers to gRPC error 
 * codes, 3-18. More details on these codes can be found [here](https://grpc.github.io/grpc/core/md_doc_statuscodes.html).
 * Note that the value of these codes is offset by 2 from their original form (e.g. `CANCELLED` = 3 vs = 1).
 * Therefore, for error codes, the transformation from gRPC to SteelEagle response code is to add
 * 2 to the code. The only codes that differ from their gRPC meaning are codes 9 and 18.
 */
enum ResponseStatus {
  OK = 0; // command acknowledged
  IN_PROGRESS = 1; // command in progress
  COMPLETED = 2; // command finished without error
  CANCELLED = 3; // the operation was cancelled, typically by the caller
  UNKNOWN = 4; // unknown error
  INVALID_ARGUMENT = 5; // the client specified an invalid argument
  DEADLINE_EXCEEDED = 6; // the deadline expired before the operation could complete
  NOT_FOUND = 7; // some requested entity was not found
  ALREADY_EXISTS = 8; // an entity the client attempted to create already exists
  PERMISSION_DENIED = 9; // the provided identity is not permitted to execute this operation by the current law (unique to SteelEagle)
  RESOURCE_EXHAUSTED = 10; // some resource has been exhausted
  FAILED_PRECONDITION = 11; // the operation was rejected because the system is not in a state required for the operation's execution
  ABORTED = 12; // the operation was aborted, typically due to a concurrency issue such as a sequencer check failure or transaction abort
  OUT_OF_RANGE = 13; // the operation was attempted past the valid range
  UNIMPLEMENTED = 14; // the operation is not implemented/supported by the service
  INTERNAL = 15; // an internal error occurred while executing the operation
  UNAVAILABLE = 16; // the service is currently unavailable
  DATA_LOSS = 17; // unrecoverable data loss or corruption
  UNAUTHENTICATED = 18; // the client failed to provide an identity (unique to SteelEagle)
}


// Global response message returned by all core services.
message Response {
  ResponseStatus status = 1; // response status
  optional string response_string = 2; // detailed message on reason for response
  google.protobuf.Timestamp timestamp = 3; // response timestamp
}


// Angular offsets or poses in 3 dimensions.
message Pose {
  optional double pitch = 1; // pitch [degrees]
  optional double roll = 2; // roll [degrees]
  optional double yaw = 3; // yaw [degrees]
}


// Representation of velocity in 3-dimensions.
message Velocity {
  optional double x_vel = 1; // forward/north velocity [meters/s]
  optional double y_vel = 2; // right/east velocity [meters/s]
  optional double z_vel = 3; // up velocity [meters/s]
  optional double angular_vel = 4; // angular velocity [degrees/s]
}


// Position offset relative to home or current location.
message Position {
  optional double x = 1; // forward/north offset [meters]
  optional double y = 2; // right/east offset [meters]
  optional double z = 3; // up offset [meters]
  optional double angle = 4; // angular offset [degrees]
}

// Location in global coordinates.
message Location {
  optional double latitude = 1; // global latitude [degrees]
  optional double longitude = 2; // global longitude [degrees]
  optional double altitude = 3; // altitude above MSL or takeoff [meters]
  optional double heading = 4; // global heading [degrees]
}
